File: core/PlatesBetweenCandles/PlatesBetweenCandles.py
from typing import List


def platesBetweenCandles(s: str, queries: List[List[int]]) -> List[int]:
    prefix_list, left_pos, right_pos = [], [], []
    count, flag, prefix_pos = 0, -1, 0
    for index, ch in enumerate(s):
        if flag != -1 and ch == "*":
            count += 1

        if ch == "|":
            flag = count
            prefix_pos = index

        left_pos.append(prefix_pos)
        prefix_list.append(0 if flag == -1 else flag)

    prefix_pos = len(s) - 1
    for index in range(len(s) - 1, -1, -1):
        if s[index] == "|":
            prefix_pos = index
        right_pos.insert(0, prefix_pos)

    res = []
    for left, right in queries:
        res.append(
            prefix_list[left_pos[right]] - prefix_list[right_pos[left]] if left_pos[right] > right_pos[left] else 0)

    return res


if __name__ == '__main__':
    print(platesBetweenCandles(s="**|**|***|", queries=[[2, 5], [5, 9]]))

File: core/PlatesBetweenCandles/PlatesBetweenCandles.java
package com.dbc;

public class PlatesBetweenCandles {
    public int[] platesBetweenCandles(String s, int[][] queries) {
        int[] left_pos = new int[s.length()];
        int[] right_pos = new int[s.length()];
        int[] prefix_sum = new int[s.length()];
        int count = 0, flag = -1, prefix_pos = 0;
        for (int i = 0; i < s.length(); i++) {
            if (flag != -1 && s.charAt(i) == '*') {
                count++;
            }
            if (s.charAt(i) == '|') {
                flag = count;
                prefix_pos = i;
            }
            prefix_sum[i] = flag;
            left_pos[i] = prefix_pos;
        }

        prefix_pos = s.length() - 1;
        for (int i = s.length() - 1; i >= 0; i--) {
            if (s.charAt(i) == '|') {
                prefix_pos = i;
            }
            right_pos[i] = prefix_pos;
        }

        int[] res = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int left = queries[i][0], right = queries[i][1];
            res[i] = right_pos[left] < left_pos[right] ? prefix_sum[left_pos[right]] - prefix_sum[right_pos[left]] : 0;
        }

        return res;
    }
}
